io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -980,28 +1077,124 @@ impl SnapshotOps for Lvol {
     fn prepare_snap_config(
         &self,
         snap_name: &str,
+        entity_id: &str,
         txn_id: &str,
     ) -> Option<SnapshotParams> {
+        // snap_name
         let snap_name = if snap_name.is_empty() {
             return None;
         } else {
             snap_name.to_string()
         };
+        let entity_id = if entity_id.is_empty() {
+            return None;
+        } else {
+            entity_id.to_string()
+        };
+
+        // txn_id
         let txn_id = if txn_id.is_empty() {
             return None;
         } else {
             txn_id.to_string()
         };
-        // Entity Id will be same as lvol uuid for the replica snapshot.
-        let entity_id = Some(self.uuid());
         // Current Lvol uuid is the parent for the snapshot.
         let parent_id = Some(self.uuid());
 
         Some(SnapshotParams::new(
-            entity_id,
+            Some(entity_id),
             parent_id,
             Some(txn_id),
             Some(snap_name),
         ))
     }
+    /// List Replica Snapshot.
+    fn list_snapshot(self) -> Vec<VolumeSnapshotDescriptor> {
+        let mut snapshot_list: Vec<VolumeSnapshotDescriptor> = Vec::new();
+        let bdev =
+            UntypedBdev::bdev_first().expect("Failed to enumerate devices");
+
+        let lvol_devices = bdev
+            .into_iter()
+            .filter(|b| b.driver() == "lvol")
+            .map(|b| Lvol::try_from(b).unwrap())
+            .collect::<Vec<Lvol>>();
+        // return empty snapshot parameter list, if no snapshot found.
+        if lvol_devices.len() <= 1 {
+            return snapshot_list;
+        }
+        for lvol in lvol_devices {
+            // skip lvol if it is not snapshot.
+            if !lvol.is_snapshot() {
+                continue;
+            }
+            let blob = lvol.bs_iter_first();
+            let mut skip_lvol = false;
+            let mut snapshot_param: SnapshotParams = Default::default();
+            for attr in SnapshotXattrs::iter() {
+                let mut val = std::ptr::null();
+                let mut size: u64 = 0;
+                let attr_id = attr.name().to_string().into_cstring();
+                let curr_attr_val;
+                unsafe {
```

Comment:
Can do the same, and just "return" `String::from_raw_parts(` from the unsafe scope.
```suggestion
                let curr_attr_val = unsafe {
```

Reply:
It is handled differently now by adding a common utility function.